### PR TITLE
Split registry FIM sync, one for keys and one for key values

### PR DESCRIPTION
**Pull request by jotacarma90**
|Related issue|
|---|
|https://github.com/wazuh/wazuh/issues/13033|

## Description
- This PR is to introduce the change related to the synchronization of FIM windows registries. 
- The FIM agent's view of registry and values tables has been removed.
- A new version of the sync messages for the manager has been added.
- Removed all functionality needed to process `:` key and key name separation.
- Now the index used is a hash formed by the fields arch, path and value_name.

## Configuration options
`    <windows_registry>HKEY_LOCAL_MACHINE\Security</windows_registry>
`
## Logs/Alerts example

Synchro message for registry keys:

`2022/02/27 18:04:05 wazuh-agent[2300] run_check.c:108 at fim_send_sync_state(): DEBUG: (6317): Sending integrity control message: {"component":"fim_registry","data":{"arch":"[x32]","attributes":{"checksum":"6853b29eef33ff39d8b63911673cf7b078f95485","gid":"0","group_name":"SYSTEM","mtime":1645882878,"perm":"{\"S-1-5-32-544\":{\"name\":\"Administradores\",\"allowed\":[\"delete\",\"read_control\",\"write_dac\",\"write_owner\",\"read_data\",\"write_data\",\"append_data\",\"read_ea\",\"write_ea\",\"execute\"]},\"S-1-5-18\":{\"name\":\"SYSTEM\",\"allowed\":[\"delete\",\"read_control\",\"write_dac\",\"write_owner\",\"read_data\",\"write_data\",\"append_data\",\"read_ea\",\"write_ea\",\"execute\"]}}","type":"registry_key","uid":"0","user_name":"Administradores"},"index":"ff03d79932df0148efa6a066552badf25ea9c466","key_path":"HKEY_LOCAL_MACHINE\\Security\\SAM\\Domains\\Account\\Aliases\\000003EB","timestamp":1645981428,"version":3},"type":"state"}`

Synchro message for key values:
`2022/02/27 18:04:05 wazuh-agent[2300] run_check.c:108 at fim_send_sync_state(): DEBUG: (6317): Sending integrity control message: {"component":"fim_value","data":{"arch":"[x32]","attributes":{"checksum":"920b517a949aec0a6fa91b0556f0a60503058fbb","hash_md5":"d41d8cd98f00b204e9800998ecf8427e","hash_sha1":"da39a3ee5e6b4b0d3255bfef95601890afd80709","hash_sha256":"e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855","size":0,"type":"registry_value","value_type":"REG_UNKNOWN"},"index":"00a7ee53218b25b5364c8773f37a38c93eae3880","key_path":"HKEY_LOCAL_MACHINE\\Security\\SAM\\Domains\\Account\\Aliases\\Names\\HelpLibraryUpdaters","timestamp":1645981428,"value_name":"","version":3},"type":"state"}`

## Tests

- Check every agent type synchronization
  - [x] 9103 branch
  - [x] 4.2.6
  - [x] 3.13
  

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
- [x] Source installation
- [x] Source upgrade
- [x] Review logs syntax and correct language
- [x] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Valgrind (memcheck and descriptor leaks check)
- Memory tests for Windows
  - [x] Scan-build report

- [x] Retrocompatibility with older Wazuh versions
- [x] Added unit tests (for new features)

### CODE REVIEW

File: src/wazuh_db/wdb_fim.c [hunk 1]
```diff
@@ -544,63 +544,64 @@ int wdb_fim_insert_entry2(wdb_t * wdb, const cJSON * data) {
...
-        if (arch == NULL) {
-            merror("DB(%s) fim/save registry request with no arch argument.", wdb->id);
-            os_free(path_escaped);
-            return -1;
-        }
-
-        if (strcmp(item_type + 9, "key") == 0) {
-            value_name = NULL;
-            full_path_length = snprintf(NULL, 0, "%s %s", arch, path_escaped);
+        if (version->valuedouble == 2.0) {
```
Comment (antoniomanuelfr) [inline src/wazuh_db/wdb_fim.c:551]: Before accessing the `valuedouble` field, we should first check if it's a number or not. We should add the following:
```c
       if (!cJSON_IsNumber(version)) {
            // Synchronization messages without the "version" attribute are ignored, but won't trigger any error
            // message.
            return 0;
        }
```

**Review by antoniomanuelfr — CHANGES_REQUESTED** [review]

File: src/wazuh_db/wdb_fim.c (hunk 1 above)
Comment (jotacarma90) [reply src/wazuh_db/wdb_fim.c:562]: Done

**Review by FrancoRivero — APPROVED** [review]
LGTM

**Review by chemamartinez — CHANGES_REQUESTED** [review]

File: src/wazuh_db/schema_upgrade_v9.sql [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+ * Copyright (C) 2015, Wazuh Inc.
+ *
+ * May 21, 2021
+ *
+ * This program is a free software, you can redistribute it
+ * and/or modify it under the terms of GPLv2.
+*/
+
+INSERT INTO sync_info (component) VALUES ('fim_registry_key');
+INSERT INTO sync_info (component) VALUES ('fim_registry_value');
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply src/wazuh_db/schema_upgrade_v9.sql:12]: Done

**Comment by jotacarma90** [comment]
## Memory usage
Reduction of memory usage during the early stages of Wazuh processing, we can observe a significant reduction with this change:

|Branch|During first scan|During first synchro|Rest of synchros|
|---|---|---|---|
|9103|35 mb|31  mb|22 mb|
|13033|35 mb|25 mb|22 mb|
|master|23 mb|18 mb|19 mb|

**Review by chemamartinez — APPROVED** [review]

**Review by MarcelKemp — APPROVED** [review]
LGTM

**Review by FrancoRivero — APPROVED** [review]
LGTM

**Review by antoniomanuelfr — CHANGES_REQUESTED** [review]
While testing the change, I have seen a bad behavior:

When registry alerts are triggered, we are still using the old mechanism when introducing the entry into the database, inserting entries using the composed `full_path` field. When the sync is triggered by the agent, these entries are replaced using the hashed index. 

We should add a new version also when raising events and introduce the entries properly into the DB

File: src/syscheckd/src/db/src/db.cpp [hunk 3]
```diff
@@ -305,6 +305,7 @@ FIMDBErrorCode fim_db_transaction_sync_row(TXN_HANDLE txn_handler, const fim_ent
             }
         }
 
+        FIMDB::instance().logFunction(LOG_ERROR, "AAAAAAA");
```
Comment (FrancoRivero) [inline src/syscheckd/src/db/src/db.cpp:308]: What is this log message? This is unclear

**Review by FrancoRivero — CHANGES_REQUESTED** [review]

File: src/unit_tests/Testing/Temporary/CTestCostData.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+---
```
Comment (FrancoRivero) [inline src/unit_tests/Testing/Temporary/CTestCostData.txt:1]: What is this file for?

File: src/unit_tests/Testing/Temporary/LastTest.log [hunk 5]
```diff
@@ -0,0 +1,3 @@
+Start testing: Apr 26 14:54 CEST
+----------------------------------------------------------
+End testing: Apr 26 14:54 CEST
```
Comment (FrancoRivero) [inline src/unit_tests/Testing/Temporary/LastTest.log:3]: What is this file for?

Comment (FrancoRivero) [reply src/unit_tests/Testing/Temporary/LastTest.log:3]: Is this really necessary?

File: src/unit_tests/Testing/Temporary/CTestCostData.txt (hunk 4 above)
Comment (FrancoRivero) [reply src/unit_tests/Testing/Temporary/CTestCostData.txt:1]: Is this really necessary?

File: src/syscheckd/src/db/src/db.cpp [hunk 6]
```diff
@@ -170,25 +170,18 @@ void fim_db_init(int storage,
                 if (sync_callback)
                 {
                     auto json = nlohmann::json::parse(msg);
+                    const auto component { json.at("component").get<std::string>() };
 
                     if (json.at("type") == "state")
```
Comment (FrancoRivero) [inline src/syscheckd/src/db/src/db.cpp:175]: If json.at("type") is different that "state", what are the other cases?

File: src/syscheckd/src/db/src/dbRegistryKey.cpp [hunk 7]
```diff
@@ -88,6 +101,8 @@ void RegistryKey::createJSON()
     data["user_name"] = m_username;
     data["group_name"] = m_groupname;
     data["mtime"] = m_time;
+    data["hash_full_path"] = m_hashpath;
+
```
Comment (FrancoRivero) [inline src/syscheckd/src/db/src/dbRegistryKey.cpp:105]: This space is unnecessary!

File: src/unit_tests/Testing/Temporary/CTestCostData.txt (hunk 4 above)
Comment (jotacarma90) [reply src/unit_tests/Testing/Temporary/CTestCostData.txt]: done

File: src/unit_tests/Testing/Temporary/LastTest.log (hunk 5 above)
Comment (jotacarma90) [reply src/unit_tests/Testing/Temporary/LastTest.log]: done

**Review by FrancoRivero — APPROVED** [review]

File: src/syscheckd/src/db/src/dbRegistryKey.cpp (hunk 7 above)
Comment (jotacarma90) [reply src/syscheckd/src/db/src/dbRegistryKey.cpp:105]: Its just for clearly reading of the code, the next line insert the json object "data" into the json object "conf"

**Review by FrancoRivero — APPROVED** [review]
LGTM

**Review by chemamartinez — APPROVED** [review]

**Review by antoniomanuelfr — APPROVED** [review]

**Review by Dwordcito — APPROVED** [review]

**Pull request merged**